table/pit-cs.go

```diff
@@ -8,56 +8,86 @@
 package table
 
 import (
-	"bytes"
-	"math/rand"
 	"time"
 
-	"github.com/cespare/xxhash"
-	"github.com/named-data/YaNFD/core"
 	"github.com/named-data/YaNFD/ndn"
 )
 
-// PitCs represents the PIT-CS tree for a thread.
-type PitCs struct {
-	root               *PitCsNode
-	ExpiringPitEntries chan *PitEntry
-
-	nPitEntries int // Number of PIT entries in tree
-	pitTokenMap map[uint32]*PitEntry
-
-	nCsEntries    int // Number of CS entries in tree
-	csReplacement CsReplacementPolicy
-	csMap         map[uint64]*CsEntry
-}
-
-// PitCsNode represents an entry in a PIT-CS tree.
-type PitCsNode struct {
-	component ndn.NameComponent
-	depth     int
-
-	parent   *PitCsNode
-	children []*PitCsNode
-
-	pitEntries []*PitEntry
-
-	csEntry *CsEntry
-}
-
-// PitEntry is an entry in a thread's PIT.
-type PitEntry struct {
-	node  *PitCsNode
-	pitCs *PitCs
-
-	Name           *ndn.Name
-	CanBePrefix    bool
-	MustBeFresh    bool
-	ForwardingHint *ndn.Name                // Interests must match in terms of Forwarding Hint to be aggregated in PIT.
-	InRecords      map[uint64]*PitInRecord  // Key is face ID
-	OutRecords     map[uint64]*PitOutRecord // Key is face ID
-	ExpirationTime time.Time
-	Satisfied      bool
-
-	Token uint32
+// PitCsTable dictates what functionality a Pit-Cs table should implement
+type PitCsTable interface {
+	InsertInterest(interest *ndn.Interest, hint *ndn.Name, inFace uint64) (PitEntry, bool)
+	RemoveInterest(pitEntry PitEntry) bool
+	// RemoveInterestByName(name *ndn.Name)
+	FindInterestExactMatch(interest *ndn.Interest) PitEntry
+	// FindInterestExactMatchByName(name *ndn.Name) PitEntry
+	FindInterestPrefixMatch(interest *ndn.Interest, token uint32) []PitEntry
+	FindInterestPrefixMatchByData(data *ndn.Data, token *uint32) []PitEntry
+	// FindInterestPrefixMatchByName(me *ndn.Name) []PitEntry
+
+	PitSize() int
+
+	// InsertOutRecord(interest *ndn.Interest, face uint64) *PitOutRecord
+	// GetOutRecords(interest *ndn.Interest) []*PitOutRecord
+	// GetOutRecordsByName(name *ndn.Name) []*PitOutRecord
+
+	InsertData(data *ndn.Data)
+	FindDataExactMatch(interest *ndn.Interest) CsEntry
+	FindDataPrefixMatch(interest *ndn.Interest) CsEntry
+	CsSize() int
+	IsCsAdmitting() bool
+	IsCsServing() bool
+
+	eraseCsDataFromReplacementStrategy(index uint64)
+
+	ExpiringPitEntries() chan PitEntry
```

Comment:
Is it normal to have a getter function for accessing a channel?

Reply:
The following code in `fw/thread.go` accesses the channel (from https://github.com/named-data/YaNFD/blob/master/fw/thread.go#L131):
```
	for !core.ShouldQuit {
		select {
		case pendingPacket := <-t.pendingInterests:
			t.processIncomingInterest(pendingPacket)
		case pendingPacket := <-t.pendingDatas:
			t.processIncomingData(pendingPacket)
		case expiringPitEntry := <-t.pitCS.ExpiringPitEntries():
			t.finalizeInterest(expiringPitEntry)
		case <-t.deadNonceList.ExpirationTimer:
			t.deadNonceList.RemoveExpiredEntry()
		case <-t.shouldQuit:
			continue
		}
	}
```

It seems like it's just a select statement and not a for-loop, so I think it should be fine? We do not range over this channel anywhere.

Reply:
OK. It seems to be a timer that sends out expiring entries. But from the name I thought it should return a list that contains all entries that are already (or going to be) expired. I think you need to explain this.

Reply:
Resolved, see https://github.com/danningyu/YaNFD/pull/2#issuecomment-1072080157.